scalyr_agent/monitor_utils/journald_checkpoints.py

```diff
@@ -0,0 +1,74 @@
+# Copyright 2014 Scalyr Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------------------
+# author:  Imron Alston <imron@scalyr.com>
+
+__author__ = 'imron@scalyr.com'
+
+import random
+import threading
+import scalyr_agent.util as scalyr_util
+import scalyr_agent.scalyr_logging as scalyr_logging
+
+global_log = scalyr_logging.getLogger(__name__)
+
+_global_lock = threading.Lock()
+_global_checkpoints = {}
+
+def load_checkpoints( filename ):
+    result = None
+    _global_lock.acquire()
+    try:
+        if filename not in _global_checkpoints:
+            checkpoints = {}
+            try:
+                checkpoints = scalyr_util.read_file_as_json( filename )
+            except:
+                global_log.log( scalyr_logging.DEBUG_LEVEL_1, "No checkpoint file '%s' exists.\n\tAll journald logs for '%s' will be read starting from their current end.", filename )
+                checkpoints = {}
+
+            result = Checkpoint( filename, checkpoints )
+            _global_checkpoints[filename] = result
```

Comment:
I'm not a big fan of holding a lock while we are doing I/O operations.  How about we hold the lock to check to see if `filename` is in `_global_checkpoints` and then again when we go to try to insert in what we just read into `_global_checkpoints`.  If someone else has added it to `_global_checkpoints` while we did the read operation, then we just drop our `Checkpoint` object on the ground and return the one in the the `_global_checkpoints` dict.

Reply:
I've fixed this up.  I agree, I'm not a big fan of locking while doing I/O either.  What about the case further down where we write the file atomically?  The atomic_write function writes to a temp file first, so I guess we'd then need to be not locking and creating unique temporary files per thread, and relying on the fact that the OS will take care of renames atomically for us without needing the lock?

I had a look at doing that initially, but the standard python tempfile mechanism didn't really have something suitable for how we use it here, and so I thought holding the lock for that I/O is probably going to be ok - at least for the journald plugin which has relatively small checkpoint files and relatively little that can go wrong during the I/O phase.  Thoughts?

Reply:
Yup, I noticed the lock being held for the writes as well... and yes, there's not as easy of a way to avoid holding it there IMO.  I prefer to not hold locks during I/O but sometimes it's just too hard to avoid it.. and yes, at least this is just on the write path so should not get much contention.